### DISCUSSION TITLE
关于armclang下smallest讨论

### DISCUSSION THREAD

**Discussion by PeakRacing** [xmake-io/xmake#5047] [General]
armclang的nf_optimize现在xmake里是直接用继承的gcc的，但是armclang有点区别。gcc是
            none       = "-O0"
        ,   fast       = "-O1"
        ,   faster     = "-O2"
        ,   fastest    = "-O3"
        ,   smallest   = "-Os"
        ,   aggressive = "-Ofast"
，armclang应该是
        none       = "-O0"
    ,   fast       = "-O1"
    ,   faster     = "-O2"
    ,   fastest    = "-O3"
    ,   smallest   = "-Oz"
    ,   aggressive = "-Ofast"
smallest对应的应该是-Oz，然后-Os armclang有一个单独的平衡模式，这个要怎么加进去呢？

另外clang下应该也是一样的，要不要再加一个optimize等级来更好的区分-Oz 和 -Os

**Comment by waruqi** [comment]
一种改成继承 clang ，一种参考 clang 的实现，单独改下 languages 提个 pr 过来

https://github.com/xmake-io/xmake/blob/c2f2ad5d2818eeefc487896d673c2585829108bf/xmake/modules/core/tools/clang.lua#L119

**Reply by PeakRacing** [reply]
这里我有看，但是还是上面的疑问，是否要再加一个新的来区别clang下os和oz

**Reply by waruqi** [reply]
不需要，不可能把所有 flags 都抽象化，这里只是抽象化一些最基本的 value，如果要更精准细力度处理，你应该直接 add_cxflags("-Oz") 而不是过于依赖 这些值在不同编译器之间的映射规则。我随时可能会改 smallest fast 等的映射规则。，对于用户不应该关系它到底映射成了什么，只要知道是 fastest 还是 smallest

**Reply by PeakRacing** [reply]
OK

**Reply by waruqi** [reply]
dev 改进过了
